Tests/UnitTests/Core/Geometry/HierarchicalGeometryContainerTests.cpp

```diff
@@ -0,0 +1,137 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2020 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include <boost/test/unit_test.hpp>
+#include <iterator>
+#include <stdexcept>
+
+#include "Acts/Geometry/HierarchicalGeometryContainer.hpp"
+
+namespace {
+
+using Acts::GeometryID;
+
+struct Thing {
+  GeometryID id;
+  double value = 1.0;
+
+  Thing(GeometryID i, double v) : id(i), value(v) {}
+
+  constexpr auto geometryId() const { return id; }
+};
+
+using Container = Acts::HierarchicalGeometryContainer<Thing>;
+
+// helper function to create geometry ids
+GeometryID makeId(int volume, int layer = 0, int sensitive = 0) {
+  return GeometryID().setVolume(volume).setLayer(layer).setSensitive(sensitive);
+}
+
+}  // namespace
+
+BOOST_TEST_DONT_PRINT_LOG_VALUE(Container::Iterator)
+
+BOOST_AUTO_TEST_SUITE(HierarchicalGeometryContainer)
+
+BOOST_AUTO_TEST_CASE(ConstructDefault) {
+  Container c;
+  BOOST_TEST(c.begin() == c.end());
+  BOOST_TEST(c.empty());
+  BOOST_TEST(c.size() == 0u);
+}
+
+BOOST_AUTO_TEST_CASE(ConstructNonUnique) {
+  std::vector<Thing> elements = {
+      {makeId(2, 4, 6), 1.0},
+      {makeId(3, 5), 1.0},
+      {makeId(3), 1.0},
+      // duplicate identifier
+      {makeId(2, 4, 6), 2.0},
+  };
+  BOOST_CHECK_THROW(Container(std::move(elements)), std::invalid_argument);
+}
+
+BOOST_AUTO_TEST_CASE(Find) {
+  Container c({
```

Comment:
@msmk0 Just to check if I understand correctly, the order of the elements in the container does matter, i.e. the finer-granularity element should be added first, right?

Reply:
The order of the input elements does not matter. The container takes ownership of the input vector and internally orders the elements to allow faster lookups.

The internal ordering is an implementation detail and should not matter to you. The logic should, i.e. if you request information for a geometry node that has no explicit entry, the first entry at higher granularity will be returned. This is independent from the ordering of the input elements.

Reply:
That's nice! Thanks! I will start to use this for the CKF.